ENH: add dependency injection point to transform X & y together

adriangb opened the pull request:
Resolves #160

adriangb commented:
@stsievert if you are able to, a review of this new interface would be much appreciated. The best place to start is probably [section 6 and 7 of this notebook](https://www.adriangb.com/scikeras/refs/pull/167/merge/notebooks/DataTransformers.html#6.-Ragged-datasets-with-tf.data.Dataset).

adriangb commented on scikeras/utils/transformers.py:
```diff
@@ -390,3 +407,31 @@ def get_metadata(self):
         }
+
+
+class ClassWeightDataTransformer(BaseEstimator, TransformerMixin):
+    """Default dataset_transformer for KerasClassifier.
+
+    This transformer implements handling of the `class_weight` parameter
+    for single output classifiers.
+    """
+
+    def __init__(self, class_weight: Optional[Union[str, Dict[int, float]]] = None):
+        self.class_weight = class_weight
+
+    def fit(
+        self,
+        data: Tuple[np.ndarray, Optional[np.ndarray], Optional[np.ndarray]],
+        dummy: None = None,
+    ) -> "ClassWeightDataTransformer":
+        return self
+
+    def transform(
+        self, data: Tuple[np.ndarray, Optional[np.ndarray], Optional[np.ndarray]]
+    ) -> Tuple[np.ndarray, Union[np.ndarray, None], Union[np.ndarray, None]]:
+        X, y, sample_weight = data
+        if self.class_weight is None or y is None:
+            return (X, y, sample_weight)
+        sample_weight = 1 if sample_weight is None else sample_weight
+        sample_weight *= compute_sample_weight(class_weight=self.class_weight, y=y)
+        return (X, y, sample_weight)
```
There is another option here, that may be cleaner, which would be to add another dependency injection point that runs in "parallel" to `feature_encoder` and `target_encoder` specifically for `sample_weight`. But since it also makes sense that one would need at least `y` to process `sample_weight`, it might be a bit redundant.

stsievert commented on docs/source/notebooks/DataTransformers.md:
```diff
@@ -43,6 +43,8 @@ In this notebook, we explore how to reconcile this functionality with the sklear
 * [5. Multidimensional inputs with MNIST dataset](#5.-Multidimensional-inputs-with-MNIST-dataset)
```
I don't know if I'd include the psuedo-code in this notebook: that seems like it should be in the API documentation, not an example.

adriangb replied:
The goal was to introduce the general structure/plan of what the notebook was going to introduce without getting bogged down in the details by introducing a fully working version.

I initially put the `if False:`s in there to avoid errors when executing, which was probably just lazy. I can probably remove all of the `if False:` by just making sure all of the pseudocode is valid python, or just marking the cell as non-executable (which is possible). Would this resolve the problem? I am not sure I am understanding how mospaced font plays into this.

stsievert replied:
> introduce the general structure/plan of what the notebook was going to introduce without getting bogged down in the details by introducing a fully working version.

I think that goal can be accomplished without any psuedo-code. "This notebook will detail how to do the following: [list]."

Marking the cell as non-executable would resolve the complaint about `if False`

adriangb replied:
I deleted most of the pseudocode and moved the rest to `advanced.rst`.

That said, for future reference, marking one of these markdown code cells as non executable is as easy as:

```
    ```python .noeval
    # code here
    ```
```

stsievert commented on docs/source/advanced.rst:
```diff
@@ -178,11 +178,50 @@ This is basically the same as calling :py:func:`~scikeras.wrappers.BaseWrapper.g
+.. code-block::
+
+                                             ↗ feature_encoder ↘
+    your data → sklearn-ecosystem → SciKeras                    dataset_transformer → Keras
```
I think this diagram is really useful.

Would this be a better diagram?


```
                                   ↗ feature_encoder ↘
    SciKeras.fit(features, labels)                    dataset_transformer → Keras.fit(dataset)
                                   ↘ target_encoder  ↗ 
```

adriangb replied:
I do like the addition of `.fit`.  My only worry is that users might think that SciKeras always creates a `tf.data.Dataset`, which is not the case, by default it gives numpy arrays to `Keras.fit`. Do you think `Keras.fit(dataset or np.array)` makes that clear? It could also be dicts of lists or something, but that's at least more uncommon.

stsievert replied:
What can `dataset_transformer` return? Only datasets/ndarrays? Or does it support the other arguments of Keras.fit?

adriangb replied:
Anything that `Keras.fit` will accept. Internally, it looks something like this:

```python3
X, y, sample_weight = dataset_transformer.fit_transform((X, y, sample_weight))
model.fit(x=X, y=y, sample_weight=sample_weight)  # aka Keras.fit
```

adriangb replied:
Maybe `Keras.fit(data)` is a good way to specify this? That way there's not confusion in interpreting `dataset` as `tf.dataDataset`. I can also add a small code block like in https://github.com/adriangb/scikeras/pull/167#discussion_r563994781 if that helps explain it.

stsievert commented on docs/source/advanced.rst:
```diff
@@ -178,11 +178,50 @@ This is basically the same as calling :py:func:`~scikeras.wrappers.BaseWrapper.g
+      to implement non-int labels (eg: strings) as well as mutli-output models.
+- Whole dataset transformations:
+    - dataset_transformer: This is the last step before passing the data to Keras.
+      It can be used to implement conversion to a `Dataset`, amongst other things.
```
```suggestion
      By default, this transformation does nothing and passes the features/labels aka `X`/`y` to [Keras.fit].
      It's signature if `dataset_transformer.fit_transform(X, y)`
      It can be used for any input that [Keras.fit] supports, including to implement conversion to a `Dataset`, or any datatype that [Keras.fit] supports.
      
[Keras.fit]:https://keras.io/api/models/model_training_apis/#fit-method
```

adriangb replied:
I ended up re-organizing this entire section, I think I incorporated this part [here](https://www.adriangb.com/scikeras/refs/pull/167/merge/advanced.html#whole-dataset-manipulation-via-data-transformer), but let me know if it is not clear.

stsievert commented on docs/source/advanced.rst:
```diff
@@ -178,11 +178,50 @@ This is basically the same as calling :py:func:`~scikeras.wrappers.BaseWrapper.g
+    - dataset_transformer: This is the last step before passing the data to Keras.
+      It can be used to implement conversion to a `Dataset`, amongst other things.
+
+`feature_encoder` and `target_encoder` are run before building the Keras Model,
```
Style nit:

```suggestion
``feature_encoder`` and ``target_encoder`` are run before building the Keras Model,
```

This is restructured text, not markdown. This is also the case elsewhere in this document.

adriangb replied:
Thanks for pointing it out. Will fix.

adriangb replied:
I _think_ I got them all.

stsievert commented on docs/source/notebooks/DataTransformers.md:
````diff
@@ -543,3 +549,297 @@ clf.fit(x_train, y_train)
+Note that `dataset_transformer` should accept a single **3 element tuple** as its argument and return value:
+
+```python
+help(KerasClassifier.dataset_transformer)
````
Style nit: link to the docs on this.

adriangb replied:
Instead of calling `help(...)` or in addition to?

stsievert replied:
I'd definitely link to the docs. 

I'm not sure calling `help` does anything meaningful. It outputs the docstring, right? I'm not sure that's super meaningful, unless you refer to a specific point later in this notebook.

adriangb replied:
I think just linking to the docstring within the docs (like [this](https://www.adriangb.com/scikeras/refs/heads/master/generated/scikeras.wrappers.KerasClassifier.html#scikeras.wrappers.KerasClassifier.feature_encoder)) should suffice. The information on the signature and such is important to understand the notebook, but I wanted to avoid duplicating the text. I'll rewrite it as a 1-2 sentence explanation in the notebook + a link to the docs.

stsievert commented on docs/source/notebooks/DataTransformers.md:
````diff
@@ -543,3 +549,297 @@ clf.fit(x_train, y_train)
+```python
+feature_1 = np.random.uniform(size=(10, ))
+feature_2 = np.random.uniform(size=(10, ))
+obs = [0] * 3 + [1] * 2 + [2] * 1 + [3] * 2 + [4] * 2
````
```suggestion
obs = [0, 0, 0, 1, 1, 2, 3, 3, 4, 4]
```

adriangb replied:
Incorporated

stsievert commented on docs/source/notebooks/DataTransformers.md:
```diff
@@ -543,3 +549,297 @@ clf.fit(x_train, y_train)
+                # Note that it is trivial to change the expected format to match Keras' ({output_name: weights, ...})
+                # see https://github.com/keras-team/keras/issues/4735#issuecomment-267473722
+                cls_wt_out = class_weight[output_num]
+                sample_weights[output_name] = compute_sample_weight(cls_wt_out, output_data)
```
Suggestion:

``` python
sample_weights = {
    compute_sample_weight(class_weight[output_num], output_data)
    for output_num, (output_name, output_data) in enumerate(zip(self.output_names, y))
}
```

adriangb replied:
Incorporated, thank you

adriangb commented on docs/source/notebooks/DataTransformers.md:
```diff
@@ -43,6 +43,8 @@ In this notebook, we explore how to reconcile this functionality with the sklear
+* [6. Ragged datasets with tf.data.Dataset](#6.-Ragged-datasets-with-tf.data.Dataset)
+* [7. Multi-output class_weight](#7.-Multi-output-class_weight)
 
 ## 1. Setup
```
I agree. I will coalesce it into `advanced.rst`.

adriangb commented on docs/source/notebooks/DataTransformers.md:
```diff
@@ -43,6 +43,8 @@ In this notebook, we explore how to reconcile this functionality with the sklear
   * [5.2 Test](#5.2-Test)
+* [6. Ragged datasets with tf.data.Dataset](#6.-Ragged-datasets-with-tf.data.Dataset)
+* [7. Multi-output class_weight](#7.-Multi-output-class_weight)
 
```
I added a similar note to the top of the notebook. Thank you for the suggestion.

adriangb commented:
@stsievert thank you for the review, the feedback was very valuable, I think I was able to incorporate most of the smaller suggestions.

For the larger stuff, I moved most of the pseudocode and general background out of the notebook and into the main docs, and added some links back to the docs instead of duplicate information, as you suggested.

Having read the documentation, do you think this `data_transformer` interface makes sense? And in particular, do you feel that using it to handle `class_weight` is problematic?

stsievert commented on docs/source/advanced.rst:
```diff
@@ -178,26 +178,126 @@ This is basically the same as calling :py:func:`~scikeras.wrappers.BaseWrapper.g
+
+Within SciKeras, this is roughly implemented as follows:
+
+.. code:: python
```
I think a lot of this code block is unnecessary. I'd use this:

``` python
class PseudoBaseWrapper:
    def fit(self, X, y, sample_weight):
        self.target_encoder_ = self.target_encoder.fit(y)
        self.feature_encoder_ = self.feature_encoder.fit(X)
        self.model_ = self._build_keras_model()
        X, y, sample_weight = self.dataset_encoder.fit(
            self.feature_encoder_.transform(X), self.target_encoder_.transform(y), sample_weight
        )
        self.model_.fit(X=X, y=y, sample_weight=sample_weight)  # tf.keras.Model.fit
        return self

    def predict(self, X):
        X = self.feature_encoder_.transform(X)
        X, _, _ = self.dataset_transformer_.fit_transform((X, None, None))
        y_pred = self.model_.predict(X)
        return self.target_encoder_.inverse_transform(y_pred)
```

adriangb replied:
I agree that we can remove the `@property` stuff. I'll implement that.

stsievert commented on docs/source/advanced.rst:
```diff
@@ -178,26 +178,126 @@ This is basically the same as calling :py:func:`~scikeras.wrappers.BaseWrapper.g
+(eg. to get the number of outputs) but *will* be able to inject data into the model building
+function (more on this below). On the other hand,
+``data_transformer`` *will* get access to the built Model, but it cannot pass any data to model building
+function.
```
> One important thing to note is that feature_encoder and target_encoder are run before building the Keras Model, while data_transformer is run after the Model is built.

This isn't clear in the example (though it's clear in the edit I've made).

> This means that the former two will not have access to the Model
(eg. to get the number of outputs) but *will* be able to inject data into the model building
function (more on this below). On the other hand,
``data_transformer`` *will* get access to the built Model, but it cannot pass any data to model building
function.

I would only say "the output of `dataset_transformer` get directly passed to tf.keras.Model.fit through self.model_.fit." All the about the number of outputs/etc is a confusing. I think I'd refer directly to the advanced examples.

adriangb replied:
> This isn't clear in the example (though it's clear in the edit I've made).

Thank you for pointing that out! The example is much better with this detail.

> I would only say

That seems reasonable, I'll cut it down to ~ what you are suggesting.

stsievert replied:
Thanks. Nothing I say hard and fast, all my comments are comments are illustrations. If a change to be hard and fast, I'll submit a PR.

stsievert commented on docs/source/advanced.rst:
```diff
@@ -229,11 +329,61 @@ type, and implements basic handling of the following cases out of the box:
+As per the table above, if you find that your target is classified as
+``"multiclass-multioutput"`` or ``"unknown"``, you will have to implement your own data processing routine.
+
+Whole dataset manipulation via data_transformer
```
I think this section should go directly below the paragraph on dataset transformers.

stsievert replied:
I'm also unsure about the level of nesting here. It looks like "multi-input" and "whole dataset" are on the same level, something I wouldn't have expected with the titles.

It might be worth collapsing this section into the "data transformers" section. This section really only mentions the signature.

adriangb replied:
I'll move it as you suggest. It then becomes duplicated with the paragraph discussed in https://github.com/adriangb/scikeras/pull/167#discussion_r565614539, which warrants removing the latter.

stsievert commented on docs/source/advanced.rst:
```diff
@@ -229,11 +329,61 @@ type, and implements basic handling of the following cases out of the box:
+This is the last step before passing the data to Keras, and it allows for the greatest
+degree of customization because SciKeras does not make any assumptions about the output data
+and passes it directly to :py:func:`tensorflow.keras.Model.fit`.
+Its signature is ``dataset_transformer.fit_transform((X, y, sample_weight))``,
```
It's signature is this:

``` python
from sklearn.base import BaseEstimator

class DatasetTransformer(BaseEstimator):
    def fit(self, X, y, sample_weight=None) -> "DatasetTransformer":
        ...
        return self

    def transform(self, X, y, sample_weight):  # return a valid input for keras.Model.fit
        ...
        return X, y, sample_weight  # option 1
        return tensorflow_dataset  # option 2
```

adriangb replied:
I may have design it wrong, and I'm not a fan of the tuple input either, but I believe it is required so that these can be chained in a pipeline.

With the signature you are proposing, this fails:

```python3
from sklearn.base import BaseEstimator
from sklearn.pipeline import make_pipeline

class DatasetTransformer(BaseEstimator):
    def fit(self, X, y, sample_weight=None) -> "DatasetTransformer":
        return self

    def transform(self, X, y, sample_weight):  # return a valid input for keras.Model.fit
        return X, y, sample_weight

p = make_pipeline(DatasetTransformer(), DatasetTransformer())

X = [1]
y = [1]

p.fit_transform(X, y)
```

If instead you accept a tuple as your input, then you can chain them:

```python3
class DatasetTransformer(BaseEstimator):
    def fit(self, X, y, sample_weight=None) -> "DatasetTransformer":
        return self

    def transform(self, data):  # return a valid input for keras.Model.fit
        X, y, sample_weight = data
        return X, y, sample_weight

p = make_pipeline(DatasetTransformer(), DatasetTransformer())

X = [1]
y = [1]

p.fit_transform((X, y, None))
```

Am I missing something?

stsievert replied:
With the example, I only meant "illustrate the signature with code, not with words." I am unsure what's happening with tuple inputs.

adriangb replied:
Oh I see. Yes, that is a good suggestion.

stsievert replied:
Maybe it'd be clearer to have the user return a dict. I'd also modify the tuple input to this transform; why not pass keyword arguments?

``` python
ValidKerasInput = Union[np.ndarray, tf.Dataset, ...]

def transform(
    self,
    X: ValidKerasInput,
    y: ValidKerasInput,
    sample_weight: Optional[ValidKerasInput]=None
) -> dict[str, ValidKerasInput]:
    ...
    return {"X": X, "y": y, "sample_weight": None}
```

adriangb replied:
That signature would work (and is ideal) for a single transformer. I think the issue is that sklearn assumes the signature `transform(X)` (i.e. a single argument), so to be compatible with the input must be a single argument. And to chain, the input and output must be identical.

That said, we could swap from a tuple to a dict. As long as it passes this test it should work:

```python3
p = make_pipeline(DatasetTransformer(), DatasetTransformer())
p.fit_transform((X, y, sample_weight)) # or {"X": X, "y"....}
```

stsievert replied:
> I think the issue is that sklearn assumes the signature transform(X) (i.e. a single argument), so to be compatible with the input must be a single argument. And to chain, the input and output must be identical.

Yes, that's the formal Scikit-learn API specification. But, that specification is not enforced. I wouldn't complain if you broke that specification: I can't imagine users will chain `DatasetTransformer`s together, and I think multiple inputs will enhance usability.

Also, conforming to API specifications is a good practice, so I won't complain if you follow the API specification.

adriangb replied:
It's not just about the API specification, in my very limited testing I already found use cases for chaining, like [this part of the notebooks](https://www.adriangb.com/scikeras/refs/pull/167/merge/notebooks/DataTransformers.html#5.-Ragged-datasets-with-tf.data.Dataset):


```python3
class RaggedClassifier(KerasClassifier):

    @property
    def dataset_transformer(self):
        t1 = FunctionTransformer(ragged_transformer)
        t2 = super().dataset_transformer  # ClassWeightDataTransformer
        t3 = FunctionTransformer(dataset_transformer)
        return make_pipeline(t1, t2, t3)
```

adriangb replied:
I was able to remove `dummy`/the second parameter by adding a `"passthrough"` step in the pipeline. This is still a quirk, but a lesser evil (I think).

stsievert replied:
I'm not sure I understand. It took a while to understand the `passthrough` parameter (I think I understand it). I can understand a tuple/dict right away, but the "passthrough" option requires some thought and digging.

I'm a fan of passing a dict around:

``` python
p = make_pipeline(DatasetTransformer(), DatasetTransformer())
data = {"X": X, "y": y, "sample_weight": sample_weight}
p.fit(data)
p.transform(data)
p.fit_transform(data)
```

I slightly prefer passing a dict around because the keys aren't fixed, and a tuple has a static length.

adriangb replied:
A dict sounds good, thank you for the suggestion. As a side note, I dug into the issue a bit more, it seems that being able to modify only X (vs. y and sample_weight like we do here) is a [known design flaw](https://github.com/scikit-learn/scikit-learn/issues/3855#issuecomment-63213164) in sklearn. This makes me feel a bit more comfortable "adapting" the API in this fashion.

Regarding the `"passthrough"`/`dummy` thing: from the [Pipeline docs](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html), the final estimator _must_ implement `fit`. And `fit` is expected to have the signature `fit(X, y)` and is called like that ([here](https://github.com/scikit-learn/scikit-learn/blob/b3ea3ed6a09fe774dfc5160a65172b1bacbb2a82/sklearn/pipeline.py#L346)). But since we are including `y` in `data`, it does not make sense to have the signature `fit(data, y=None)` and then override `y` in the body of the function. Hence why I originally renamed the parameter from `y` to `dummy` to avoid confusion. The alternative, which is a fix specifically for pipelines but may leave room for issues elsewhere, is to use the better signature of `fit(data)` but force the last estimator in the pipeline to be `"passthrough"`/`None`. This means that all of these transformers are "known" to be transformers and so they are called with just `X` (here).

I'm a bit torn here, on the one hand, I don't like having a parameter called `dummy`. On the other hand, I think the `"passthrough"` solution is more fragile (and it isn't very nice either).

stsievert replied:
> ` fit(data, y=None)`

Why not `fit(data, **ignored)` or `fit(data, **empty)`? If you want, you can even write `assert len(empty) == 0`.

adriangb replied:
I agree. Let me try implementing this tomorrow.

For the dict, do you envision passing/returning `None` or just not including the key?

Also, a minor nit, but sklearn uses `X` (capital) while Keras uses `x` (lower). There's something to be said for using `x`, since it allows for `self.model_.fit(**data, ...)`. Do you have a preference for `X` or `x`?

stsievert replied:
I think it's easiest to say "the keys in the returned dictionary should be valid keywords to keras.Model.fit".

Let's take a transformer that does nothing and returns the input, in-modified.

Then, `x` should be preferred because it's a valid keyword argument to keras.Model.fit. I suspect that answers the no key question.

adriangb replied:
If we are going to go with dicts, what are your thoughts on passing all `Model.fit`/`Model.predict` `**kwargs`?

The pro is that it would allow nearly arbitrary user logic injection. For example, #131 could be resolved by creating a transformer that siphons off validation data from `x` & `y` into `validation_dataset`. This would be done after applying target encoding, etc. so it should be compatible with all of sklearn and Keras.

The con is that I think the interface could be messy.

stsievert replied:
How would it be messy? Arbitrary user logic sounds like a good feature for something basic like this.

> If we are going to go with dicts, what are your thoughts on passing all Model.fit/Model.predict **kwargs?

Doesn't a user-passed kwarg already passed to `fit`? I'm thinking of `scikeras_model.fit(...,  batch_size=42)`. That's in `kwargs` and gets passed to `model_.fit`, right?

adriangb replied:
> How would it be messy?

It's minor, but it would now be a dict with dozens of keys, some values being arrays, etc. It can easily be documented, but I can see how at first glance it might be more confusing than a 3 element tuple or a 3 key dict.

> Doesn't a user-passed kwarg already passed to fit?

Yes, but these parameters can't be calculated based on the data, which becomes especially problematic if doing cross-valdation and such. This is the basic problem in #131.

Of course, if users want to pass a fixed parameter like `batch_size=42` they can do so via `fit` (or better yet, via the constructor).

stsievert replied:
> It's minor, but it would now be a dict with dozens of keys,

Why would it have to be dozens of keys? I'm thinking of this code:

``` python
# line 885 of wrappers.py
transform_kwargs = self.dataset_transformer_.transform(x=X, y=y, sample_weight=sample_weight)
kwargs.update(transform_kwargs)
self._fit_keras_model(**kwargs)
```

adriangb replied:
I think that implementation in https://github.com/adriangb/scikeras/pull/167#discussion_r576512695 is not compatible with the sklearn ecosystem, including pipelines, which I think are in important feature. This would be fine:

```python3
self.dataset_transformer_.transform(dict(x=X, y=y, sample_weight=sample_weight))
```

Regarding the multiple keys, what I was thinking was to pass _all_ `Model.fit` keys. In `_fit_keras_model`:

https://github.com/adriangb/scikeras/blob/6eee3c4a19b28694758e02aab27af9222ff82db8/scikeras/wrappers.py#L480-L494

This means that users can not only add keys, but also modify existing keys, to their liking.

Regarding the `dozens of keys` comment, I wasn't referring to the keys assigned via `kwargs.update(transform_kwargs)` above, I was more thinking that when a user creates a transformer like:

```python
def transform(data: Dict[str, Any) -> Dict[str, Any]:
    len(data.keys())  # many keys, array-values, etc.
    ...

# for context:
class SomeCustomEstimator(BaseWrapper):
    @property
    def data_transformer(self):
        return FunctionTransformer(transformer)
```

stsievert replied:
> This means that users can not only add keys, but also modify existing keys, to their liking.

Yes, I think that's a good idea. It also leads to simple documentation (something like "the dataset transformer is passed a dict. Normally, this dictionary is unmodified and passed directly to keras.Model.fit. If desired, it's possible to modify this dictionary through a Scikit-learn transformation: [example]."

adriangb replied:
Great. I'm going to loop back to #131 and #159 to see if the issue reporters have any input on this implementation.

stsievert commented on scikeras/wrappers.py:
```diff
@@ -755,20 +797,24 @@ def _initialize(
-        return X, y
+        self.dataset_transformer_ = self.dataset_transformer.fit((X, y, sample_weight))
+        dataset_meta = getattr(self.dataset_transformer_, "get_metadata", dict)()
+        vars(self).update(**dataset_meta)
```
:-1:

stsievert replied:
```suggestion
        self.dataset_meta_ = dataset_meta
```

adriangb replied:
Which part of these 3 lines are you 👎 on? This code hasn't stayed up to date with our discussions, so I'm not sure if it's the tuple part or something else.

adriangb replied:
Ah, sorry, I see now we cross posted.

stsievert replied:
I think this comment is tangential to the other discussion. I would rather consolidate similar keys into one dictionary. I'm not a fan of blindly assigning attributes to objects.

adriangb replied:
Agreed. We actually weren't using this at the moment, and there's no reason it can't be added at a later date if a good feature requires it, so I was able to remove it for now. Does that sound like it resolves this problem?

stsievert replied:
Yeah, I think so. I'm not opposed to setting the metadata about the dataset in a single attribute:

``` python
self.dataset_meta_ = getattr(self.dataset_transformer_, "get_metadata", dict)()
```

adriangb replied:
Since I removed calling `get_metadata` for `dataset_transformer_`, I'm going to resolve this comment.
